### PR TITLE
Refactor(plugins): Unify python dynamic loader

### PR DESCRIPTION
**Pull request by gmuloc**
## Related Issue(s)

Fixes #<ISSUE ID>

## Component(s) name

`arista.avd.<role-name>`

### User Checklist

- N/A

### Repository Checklist

- [ ] My code has been rebased from devel before I start
- [ ] I have read the [**CONTRIBUTING**](https://avd.sh/en/latest/docs/contribution/overview.html) document.
- [ ] My change requires a change to the documentation and documentation have been updated accordingly.
- [ ] I have updated [molecule CI](https://github.com/aristanetworks/ansible-avd/tree/devel/ansible_collections/arista/avd/molecule) testing accordingly. (check the box if not applicable)

### CODE REVIEW

**Comment by ClausHolbechArista** [comment]
please rebase

File: ansible_collections/arista/avd/plugins/action/yaml_templates_to_facts.py [hunk 1]
```diff
@@ -135,26 +134,18 @@ def run(self, tmp=None, task_vars=None):
...
-                        try:
-                            template_result_data = cls_instance.render()
-                        except Exception as error:
-                            raise AnsibleActionFail(error) from error
-                    else:
-                        raise AristaAvdError(f"{cls_instance} has no attribute render")
+                cls_instance = cls(hostvars=template_vars, templar=templar)
+                if debug:
+                    debug_item["timestamps"]["render_python_class"] = datetime.now()
+                if getattr(cls_instance, "render"):
```
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/plugins/action/yaml_templates_to_facts.py]: this check should not be reversed to raise inside if (avoid indentation of the main action)

File: ansible_collections/arista/avd/plugins/plugin_utils/utils/load_python_class.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
+import importlib
+from typing import Optional
+
+from ansible_collections.arista.avd.plugins.plugin_utils.errors import AristaAvdError
+
+
+def load_python_class(module_path: str, class_name: str, parent_class: Optional[type] = None) -> type:
```
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/plugins/plugin_utils/utils/load_python_class.py]: We don't use Optional anywhere else. I'm not sure we need it here. Could be `: type | None`

File: ansible_collections/arista/avd/plugins/plugin_utils/utils/load_python_class.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+    Raises
+    ------
+    AristaAvdError
+        If module_path or class_name are not present
+        If importlib fails to load the Class
+        If the loaded Class is not inheriting from the optional parent_class
+    """
+    if not module_path or not class_name:
+        raise AristaAvdError("Cannot load a class from a module if class_name or module_path are not given")
```
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/plugins/plugin_utils/utils/load_python_class.py]: This is not clear in the context of someone forgetting to set the module under `templates` yaml. Maybe include the information we have to guide them.

File: ansible_collections/arista/avd/roles/eos_designs/python_modules/interface_descriptions/__init__.py [hunk 4]
```diff
@@ -1,4 +1,33 @@
...
 from .avdinterfacedescriptions import AvdInterfaceDescriptions
-from .load_interface_descriptions import load_interfacedescriptions
 
 __all__ = ["AvdInterfaceDescriptions", "load_interfacedescriptions"]
+
+DEFAULT_AVD_INTERFACE_DESCRIPTIONS_PYTHON_MODULE = "ansible_collections.arista.avd.roles.eos_designs.python_modules.interface_descriptions"
+DEFAULT_AVD_INTERFACE_DESCRIPTIONS_PYTHON_CLASS_NAME = "AvdInterfaceDescriptions"
+
+
+def load_interfacedescriptions(hostvars, templar) -> AvdInterfaceDescriptions:
```
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/roles/eos_designs/python_modules/interface_descriptions/__init__.py]: this should be in it's own file. Same for ip

File: ansible_collections/arista/avd/plugins/plugin_utils/utils/load_python_class.py (hunk 2 above)
Comment (gmuloc) [reply ansible_collections/arista/avd/plugins/plugin_utils/utils/load_python_class.py]: https://mypy.readthedocs.io/en/stable/command_line.html#none-and-optional-handling

https://peps.python.org/pep-0484/#union-types

_This is no longer the recommended behavior. Type checkers should move towards requiring the optional type to be made explicit._

File: ansible_collections/arista/avd/roles/eos_designs/python_modules/interface_descriptions/__init__.py (hunk 4 above)
Comment (gmuloc) [reply ansible_collections/arista/avd/roles/eos_designs/python_modules/interface_descriptions/__init__.py]: that makes a file to maintain for one function but will change this!

File: ansible_collections/arista/avd/plugins/plugin_utils/utils/load_python_class.py (hunk 3 above)
Comment (gmuloc) [reply ansible_collections/arista/avd/plugins/plugin_utils/utils/load_python_class.py]: inside the method I don't have much information actually I don't know if we are inside templates or loader but will make this better

File: ansible_collections/arista/avd/plugins/action/yaml_templates_to_facts.py [hunk 5]
```diff
@@ -134,18 +134,23 @@ def run(self, tmp=None, task_vars=None):
             elif "python_module" in template_item:
                 module_path = template_item.get("python_module")
                 class_name = template_item.get("python_class_name", DEFAULT_PYTHON_CLASS_NAME)
-                cls = load_python_class(module_path, class_name, AvdFacts)
+                try:
+                    cls = load_python_class(module_path, class_name, AvdFacts)
+                except AristaAvdMissingVariableError as exc:
+                    raise AristaAvdError(f"Missing module_path or class_name in {template_item}") from exc
```
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/plugins/action/yaml_templates_to_facts.py]: Should really raise AnsibleActionFail

File: ansible_collections/arista/avd/plugins/action/yaml_templates_to_facts.py [hunk 6]
```diff
@@ -134,18 +134,23 @@ def run(self, tmp=None, task_vars=None):
...
                     debug_item["timestamps"]["render_python_class"] = datetime.now()
-                if getattr(cls_instance, "render"):
-                    try:
-                        template_result_data = cls_instance.render()
-                    except Exception as error:
-                        raise AnsibleActionFail(error) from error
-                else:
+
+                if not (getattr(cls_instance, "render")):
                     raise AristaAvdError(f"{cls_instance} has no attribute render")
```
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/plugins/action/yaml_templates_to_facts.py]: Should really raise AnsibleActionFail

**Pull request merged**